docs/DafnyRef/Types.md

```diff
@@ -1172,21 +1203,248 @@ can be used abstractly to represent an arbitrary parameterized monad.
+  "="
+  LocalIdentTypeOptional
+  "|"
+  Expression(allowLemma: false, allowLambda: true)
+  [ "ghost" "witness" Expression(allowLemma: false, allowLambda: true)
+  | "witness" Expression((allowLemma: false, allowLambda: true)
+  | "witness" "*"
+  ]
+````
+
+````grammar
+NatType_ = "nat"
+````
+
+A _subset type_ is a restricted use of an existing type, called
+the _base type_ of the subset type.  A subset type is like a
+combined use of the base type and a predicate on the base
+type.
+
+An assignment from a subset type to its base type is always
+allowed.  An assignment in the other direction, from the base type to
+a subset type, is allowed provided the value assigned does indeed
+satisfy the predicate of the subset type. This condition is checked
+by the verifier, not by the type checker. Similarly, assignments from
+one subset type to another (both with the same base type) are also
+permitted, as long as it can be established that the value being assigned
+satisfies the predicate defining the receiving subset type.
+(Note, in contrast, assignments between a newtype and its base type
+are never allowed, even if the value assigned is a value of the target
+type.  For such assignments, an explicit conversion must be used, see
+[Section 20.10](#sec-as-expression).)
+
+Dafny supports a built-in subset type, namely the type `nat`,
+whose base type is `int`. Type `nat`
+designates the non-negative subrange of `int`.  A simple example that
+puts subset type `nat` to good use is the standard Fibonacci
+function:
+```dafny
+function Fib(n: nat): nat
+{
+  if n < 2 then n else Fib(n-2) + Fib(n-1)
+}
+```
+An equivalent, but clumsy, formulation of this function (modulo the
+wording of any error messages produced at call sites) would be to use
+type `int` and to write the restricting predicate in pre- and
+postconditions:
+```dafny
+function Fib(n: int): int
+  requires 0 <= n  // the function argument must be non-negative
+  ensures 0 <= Fib(n)  // the function result is non-negative
+{
+  if n < 2 then n else Fib(n-2) + Fib(n-1)
+}
+```
+
+TODO: I think the following is false, or at least should be.
+
+Type inference will never infer the type of a variable to be a
+subset type.  It will instead infer the type to be the base type
+of the subset type.  For example, the type of `x` in
+```dafny
+forall x :: P(x)
+```
+will be `int`, even if predicate `P` declares its argument to have
+type `nat`.
+
 <!--PDF NEWPAGE-->
-# 12. Class Types {#sec-class-types}
+# 12. Newtypes {#sec-newtypes}
+````grammar
+NewtypeDecl = "newtype" { Attribute } NewtypeName "="
+  [ ellipsis ]
+  ( LocalIdentTypeOptional
+    "|"
+    Expression(allowLemma: false, allowLambda: true)
+    [ "ghost" "witness" Expression(allowLemma: false, allowLambda: true)
+    | "witness" Expression((allowLemma: false, allowLambda: true)
+    | "witness" "*"
+    ]
+  | Type
+  )
+  [ TypeMembers ]
+````
+A newtype is like a type synonym except that it declares a wholly new type
```

Comment:
```suggestion
A newtype is like a type synonym or subset type except that it declares a wholly new type
```

Reply:
OK, though I think of a subset type as one kind of type synonym